backend/docker.go

```diff
@@ -312,36 +323,121 @@ func (i *dockerInstance) sshClient() (*ssh.Client, error) {
 }
 
 func (i *dockerInstance) UploadScript(ctx gocontext.Context, script []byte) error {
-	client, err := i.sshClient()
+	tarBuf := &bytes.Buffer{}
+	tw := tar.NewWriter(tarBuf)
+	err := tw.WriteHeader(&tar.Header{
+		// how to not hardcode?
+		Name: "/home/travis/build.sh",
+		Mode: 0755,
+		Size: int64(len(script)),
+	})
 	if err != nil {
 		return err
 	}
-	defer client.Close()
-
-	sftp, err := sftp.NewClient(client)
+	_, err = tw.Write(script)
+	if err != nil {
+		return err
+	}
+	err = tw.Close()
 	if err != nil {
 		return err
 	}
-	defer sftp.Close()
 
-	_, err = sftp.Lstat("build.sh")
-	if err == nil {
-		return ErrStaleVM
+	uploadOpts := docker.UploadToContainerOptions{
+		InputStream: bytes.NewReader(tarBuf.Bytes()),
+		Path:        "/",
 	}
 
-	f, err := sftp.Create("build.sh")
+	// fmt.Printf("---> Uploading to container: %#v\n", uploadOpts)
+
+	return i.client.UploadToContainer(i.container.ID, uploadOpts)
+}
+
+func (i *dockerInstance) RunScript(ctx gocontext.Context, output io.Writer) (*RunResult, error) {
+	if os.Getenv("TRAVIS_WORKER_DOCKER_EXEC") == "1" {
+		return i.runScriptExec(ctx, output)
+	}
+	return i.runScriptSSH(ctx, output)
+}
+
+func (i *dockerInstance) runScriptExec(ctx gocontext.Context, output io.Writer) (*RunResult, error) {
+	createExecOpts := docker.CreateExecOptions{
+		// TODO: remove hardcoded shenanigans
+		AttachStdin:  false,
+		AttachStdout: true,
+		AttachStderr: true,
+		Tty:          true,
+		Cmd:          []string{"bash", "-l", "-c", "cd /home/travis && exec bash build.sh"},
```

Comment:
I think the build script expects to be run in `/home/travis`, and I think maybe `docker exec` has `/` as the working directory by default? (I guess a `docker exec pwd` or `docker exec ls` can probably tell you whether that is right)